runtime/connectors/blob/objectreader.go

```diff
@@ -0,0 +1,94 @@
+package blob
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	"github.com/c2h5oh/datasize"
+	"gocloud.dev/blob"
+)
+
+// ObjectReader reads range of bytes from cloud objects
+// implements io.ReaderAt and io.Seeker interfaces
+type ObjectReader struct {
+	ctx    context.Context
+	bucket *blob.Bucket
+	index  int64
+	obj    *blob.ListObject
+
+	// debug data
+	debugMode bool
+	call      int64
+	bytes     int64
+}
+
+func (f *ObjectReader) ReadAt(p []byte, off int64) (int, error) {
+	if f.debugMode {
+		fmt.Printf("reading %v bytes at offset %v\n", len(p), off)
+		atomic.AddInt64(&f.call, 1)
+	}
```

Comment:
I am not aware of go in particular but this kind of metric collection is very common. Also every log statement should be running a similar check internally. If configured log level is higher than request log level don't log anything.
I can see multiple use cases for this like : 
1. Collecting sampled metrics
2. In case users complaint about cost, this flag can be enabled to check how many calls does our system make (similar to what duckdb's explain analyse do).
3. Reporting these stats to users if queried. 

Ideally this should be controlled by some env variable or some flag passed to cli and we should be using that variable in entire code but I think that can be tackled separately. 

The performance implications are definitely trivial unless debugMode is enabled since its just executing an if condition. 
Some implementations that I have seen (in other languages) use a stats interface that exposes functions to change internal counters. 
In debug mode the stats impl actually increment some counters.
In normal mode the stats impl is a dummy object with no counters and does nothing on function calls.
That felt like an overkill here. In case we want to go ahead with that approach I can add that as well. 

Happy to discuss any other reservations that you have in keeping this.

Reply:
Okay, that makes sense, we can keep it for now. We want to introduce proper metrics collection at some point (across all backend areas), we can revisit when we prioritize that.